File: src/leetcode/tree/993_Cousins_in_Binary_Tree.java
//
// Definition for a binary tree node.
// public class TreeNode {
//    int val;
//     TreeNode left;
//    TreeNode right;
//    TreeNode(int x) { val = x; }
//}
//
class Solution {
    Map<Integer, Integer> depths;
    Map<Integer, TreeNode> parents;

    public boolean isCousins(TreeNode root, int x, int y) {
        depths = new HashMap();
        parents = new HashMap();
        dfs(root, null);
        return depths.get(x) == depths.get(y) && parents.get(x) != parents.get(y);
    }

    public void dfs(TreeNode node, TreeNode parent) {
        if (node == null) {
            return;
        }
        depths.put(node.val, parent != null ? depths.get(parent.val) + 1 : 0);
        parents.put(node.val, parent);
        dfs(node.left, node);
        dfs(node.right, node);
    }
}

File: src/leetcode/tree/993_Cousins_in_Binary_Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} x
 * @param {number} y
 * @return {boolean}
 */
var isCousins = function (root, x, y) {
    var queue = [];
    queue.push(root);
    var a = false,
        b = false;
    while (queue.length) {
        var size = queue.length;
        if (a !== b) {
            return false;
        }
        for (var i = 0; i < size; i++) {
            var cur = queue.shift();
            if (cur.val === x) {
                a = true;
            }
            if (cur.val === y) {
                b = true;
            }
            if (cur.left !== null && cur.right !== null) {
                if (cur.left.val === x && cur.right.val === y) {
                    return false;
                }
                if (cur.right.val === x && cur.left.val === y) {
                    return false
                }
            }
            if (cur.left !== null) {
                queue.push(cur.left);
            }
            if (cur.right !== null) {
                queue.push(cur.right);
            }
        }
        if (a && b) {
            return true;
        }
    }
    return false;
};